---
title: "Class 09: Structural Bioinformatics 1"
author: "Ariane"
format: pdf
editor_options: 
  chunk_output_type: inline
---

# What is the PDB anyway?

The main database of biomolecular structures is called the PDB and is available at www.rcsb.org.

What is in this database:

```{r}
pdbstats <- read.csv("PDB.csv", row.names = 1)
head(pdbstats)
```

Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
n.xray <- sum(as.numeric(gsub(",",'', pdbstats$X.ray)))
n.em <- sum(as.numeric(gsub(",",'', pdbstats$EM))) 
n.total <- sum(as.numeric(gsub(",",'', pdbstats$Total))) 

p.xray <- (n.xray/n.total) * 100
p.em <- (n.em/n.total) * 100

round(p.xray, 2)
round(p.em, 2)
```

There are `r round(n.xray)` protein structures (`r round(p.xray, 2)`%) and `r round(n.em)` (`r round(p.em, 2)`%) EM structures in the current PDB database. 

Q2: What proportion of structures in the PDB are protein?

```{r}
as.numeric(gsub(",",'', pdbstats$Total)) / n.total
```


Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

It is not straight forward to find all HIV-1 protease structures using pain text searching on the database. 

Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

The hydrogen atom is too small to be visible at this resolution. Therefore, we can only observe the oxygen atom in each water molecule in this structure. 

Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

HOH 308

Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?

A pic of HIV-1 Protease from Molstar

![An image I like while learning how to break Molstar](1HSG.png)

# Working with structure data in R

We will use the `bio3d` package for this:

```{r}
library(bio3d)
```

RBD 
```{r}
pdb <- read.pdb("1hsg")
pdb
```

```{r}
head(pdb$atom)
```

What is the first residue 3 letter code?

```{r}
pdb$atom$resid[1]
aa321(pdb$atom$resid[1])
```

Q7: How many amino acid residues are there in this pdb object? 

198

Q8: Name one of the two non-protein residues? 

HOH

Q9: How many protein chains are in this structure?

2

```{r}
attributes(pdb)
```

```{r}
head(pdb$atom)
```

# Prediciting functional motions of a single structure 

Let’s read a new PDB structure of Adenylate Kinase and perform Normal mode analysis.

```{r}
adk <- read.pdb("6s36")
adk
```

```{r}
m <- nma(adk)
plot(m)
```

```{r}
mktrj(m, file="adk_m7.pdb")
```

Q10. The package `msa` is only found on BioConductor not CRAN. 

Q11. The package `Grantlab/bio3d-view` is not found on either BioConductor or CRAN. 

Q12. It is true that the functions from the devtools package can be used to install packages from GitHub and BitBuckest. 

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

Q13. 214 amino acids in this sequence.

```{r}
#b <- blast.pdb(aa)
```

to render it without running blast each time but still have the "b" object, we can save the b object and load it next time. 

```{r}
#saveRDS(b, file ="blast_1ake_A.RDS")
b <- readRDS("blast_1ake_A.RDS")
```

```{r}
# Plot a summary of search results
hits <- plot(b)
```

```{r}
# List out some 'top hits'
head(hits$pdb.id)
```

```{r}
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
#plot(pdbs, labels=ids)
```

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
anno
```

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

```{r}
# NMA of all structures
modes <- nma(pdbs)
```
```{r}
plot(modes, pdbs, col=grps.rd)
```

Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

The fluctuation of the black line is lower than the colored lines. The black and colored lines look different. The regions between 25-60 and 125-160 differ the most. 